doc/ztp/ztp_hld.md

```diff
@@ -0,0 +1,166 @@
+    - Connect 
+    - Power-on 
+2.	Onie Boots into SONiC (with ZTP enabled) from flash
+    - Power-on ZTP service will take over to load the initial configuration.
+    - ZTP service flag $enabled is true
+    - ZTP service flag $post_install is false
+    - ZTP service flag $post_config is false
+3.	The switch now enters ZTP mode and does the following,
+    - Obtains an IP address from the DHCP server
+    - Obtains the URL for the SONIC image, minigraph, and post config validation script 
+4.	At this step, the switch will have information to reach the HTTP / TFTP server through URLs given by DHCP server.
+5.	ZTP service will enable and start updategraph service
+    - Within updategraph, ztp_enabled is true, and post_install is false
+    - This triggers updategraph to configure the device with an default config
+    - Upon exiting updategraph, SNMP will start
+6.	ZTP service will now download the software image file.
+7.	sonic_installer will validate image to be a SONIC image compatible with the current platform.
+8.	sonic_installer  will check remaining hard disk for space before doing image install.
+9.	ZTP service will install image using sonic-installer
+    - ZTP service flag $post_install is true
+10.	ZTP service will enable updategraph
+11.	ZTP service will call for reboot to apply new SONIC image, with option ZTP-enabled –post_install
+12.	Upon reboot, if ZTP service flag $post_install is true, ZTP hands off to updategraph
+13.	The configuration script is received and applied by updategraph, returns to ZTP service
+ZTP service flag $post_config is true
+14.	Validation:
+    - SONIC device will download post config validation script through ZTP enabled DHCP option
+    - The script is customized based on user preference
+    - The script collects local info and sends it back to server using information in the script itself
+15.	Delete old SONIC image upon successful reimage/ ZTP status check:
+    - Check if $post_install is true
+    - Check if $post_config is true
+    - Set $enable to false
+    - Verify no errors were thrown during the service
+    - Output to syslog server and var/log/syslog
+
+### Updategraph
+Needs to be updated to provide default config to the switch when ztp_enabled is true
+Needs to continue as currently designed when ztp_enabled is true and post_install is true
+Can be used to download config file and apply config
+
+### Interruption of ZTP
+Through command line utility “ZTP Disable” 
+Sets ZTP $enabled flag to FALSE
+If ZTP is interrupted mid-service, output is shown
+If ZTP service receives “NA_ZTP” flag, continue updategraph with default config
+ZTP service needs to have separate option for interruption than updategraph
+If ZTP service is interrupted, updategraph should still be enabled and allowed to continue
+
+### Syslog
+ZTP status should be logged to syslog server and /var/log/syslog
+
+### Build Option
+Need to allow build time option to compile with ZTP enabled
+
+# Error Cases
```

Comment:
Should probably add some cases for network events, such as: -
- DHCP server does not respond
- URL targets not reachable, or lost connection during load
- etc